### PR TITLE
Add support for property dependency injection

### PR DESCRIPTION
**Pull request by raymondfeng**
cc @bajtos @ritch @superkhau

### CODE REVIEW

File: packages/context/test/unit/inject.test.ts [hunk 1]
```diff
@@ -22,7 +22,8 @@ describe('function argument injection', () => {
     }
 
     const meta = describeInjectedArguments(TestClass);
-    expect(meta).to.deepEqual(['foo']);
+    expect(meta.length).to.eql(1);
+    expect(meta[0].bindingKey).to.deepEqual('foo');
```
Comment (bajtos) [inline packages/context/test/unit/inject.test.ts]:
```ts
expect(meta.map(m => m.bindingKey)).to.deepEqual(['foo']);
```

This way the assertion error message can include extra keys when the meta has more than one entry.

File: packages/context/test/unit/resolver.test.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    ctx.bind('foo').to('FOO');
+  });
+
+  it('can resolve constructor arguments', () => {
+    class TestClass {
+      constructor(@inject('foo') public foo: string) {
+      }
+    }
+
+    let t = instantiateClass<TestClass>(TestClass, ctx) as TestClass;
```
Comment (bajtos) [inline packages/context/test/unit/resolver.test.ts]: I think the TypeScript compiler should be able to infer the specialisation of `instantiateClass` from the first argument:

```ts
let t = instantiateClass(TestClass, ctx) as TestClass;
``` 

Could you please double check?

If I am right, then the same simplification should be applied to most of the other tests below too.

Comment (bajtos) [inline packages/context/test/unit/resolver.test.ts]: Also we should prefer `const` over `let`, I am surprised `tslint` did not complain.

File: packages/context/test/unit/resolver.test.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    expect(t.foo).to.eql('FOO');
+
```
Comment (bajtos) [inline packages/context/test/unit/resolver.test.ts]: Extra empty line, please remove.

Same comment applies to other test cases in this file too.

**Comment by bajtos** [comment]
> Also we should prefer `const` over `let`, I am surprised `tslint` did not complain.

Haha, tslint did catch that, see the output of Travis CI for an example ([link](https://travis-ci.org/strongloop/loopback-next/jobs/235034469)). Please fix your code to pass`npm test` ;-)

File: packages/context/test/unit/resolver.test.ts (hunk 2 above)
Comment (raymondfeng) [inline packages/context/test/unit/resolver.test.ts]: Please note the return type of `instantiateClass` is `T | Promise<T>`. We have to cast it.

**Comment by raymondfeng** [comment]
@bajtos Comments addressed. PTAL.

File: packages/context/src/inject.ts [hunk 4]
```diff
@@ -26,19 +35,48 @@ const REFLECTION_KEY = 'loopback.inject';
...
+ * @param metadata Optional metadata to help the injection
+ *
  */
-export function inject(bindingKey: string) {
-  return function markArgumentAsInjected(target: Object, propertyKey: string | symbol, parameterIndex: number) {
-    assert(parameterIndex != undefined, '@inject decorator can be used on function arguments only!');
+export function inject(bindingKey: string, metadata?: Object) {
+  // tslint:disable-next-line:no-any
+  return function markArgumentAsInjected(target: any, propertyKey?: string | symbol,
+    propertyDescritorOrParameterIndex?: TypedPropertyDescriptor<BoundValue> | number) {
```
Comment (bajtos) [inline packages/context/src/inject.ts]: Typo?

```diff
- propertyDescritorOrParameterIndex
+ propertyDescriptorOrParameterIndex
```

File: packages/context/src/resolver.ts [hunk 5]
```diff
@@ -23,10 +23,34 @@ export type Constructor<T> = new(...args: any[]) => T;
...
+      } else {
+        // Inject the properties synchrounously
+        Object.assign(inst, props);
+        return inst;
+      }
+    });
+  } else {
+    if (isPromise(inst)) {
+      // Inject the properties asynchrounously
+      return inst.then(obj => Object.assign(obj, propertiesOrPromise));
```
Comment (bajtos) [inline packages/context/src/resolver.ts]: This is inconsistent with L39. Why is `return obj` needed on L39 but not needed here?

Similarly for L43, why is `return inst` needed?

File: packages/context/test/unit/resolver.test.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    ctx.bind('foo').to('FOO');
+  });
+
+  it('can resolve constructor arguments', async () => {
+    class TestClass {
+      constructor(@inject('foo') public foo: string) {
+      }
+    }
+
+    const t = await instantiateClass(TestClass, ctx) as TestClass;
```
Comment (bajtos) [inline packages/context/test/unit/resolver.test.ts]: On L123 below, you have the following code:

```ts
const t = await instantiateClass(TestClass, ctx);
```

Can we use the same approach here and avoid the extra `as TestClass` cast?

Please review other tests calling `await instantiateClass` too and make the code consistent.

**Comment by raymondfeng** [comment]
@bajtos Fixed. PTAL.

**Pull request merged**